pkg/model/utxo/receipt.go

```diff
@@ -0,0 +1,188 @@
+package utxo
+
+import (
+	"encoding/binary"
+
+	"github.com/gohornet/hornet/pkg/model/hornet"
+	"github.com/gohornet/hornet/pkg/model/milestone"
+	"github.com/iotaledger/hive.go/kvstore"
+	"github.com/iotaledger/hive.go/marshalutil"
+	iotago "github.com/iotaledger/iota.go/v2"
+)
+
+// ReceiptTuple contains a receipt and the index of the milestone
+// which contained the receipt.
+type ReceiptTuple struct {
+	// The actual receipt.
+	Receipt *iotago.Receipt
+	// The index of the milestone which included the receipt.
+	MilestoneIndex milestone.Index
+}
+
+func (rt *ReceiptTuple) kvStorableKey() (key []byte) {
+	return marshalutil.New(9).
+		WriteByte(UTXOStoreKeyPrefixReceipts).
+		WriteUint32(rt.Receipt.MigratedAt).
+		WriteUint32(uint32(rt.MilestoneIndex)).
+		Bytes()
+}
+
+func (rt *ReceiptTuple) kvStorableValue() (value []byte) {
+	receiptBytes, err := rt.Receipt.Serialize(iotago.DeSeriModeNoValidation)
+	if err != nil {
+		panic(err)
+	}
+	return receiptBytes
+}
+
+func (rt *ReceiptTuple) kvStorableLoad(_ *Manager, key []byte, value []byte) error {
+	keyExt := marshalutil.New(key)
+
+	// skip prefix and migrated at index
+	if _, err := keyExt.ReadByte(); err != nil {
+		return err
+	}
+
+	if _, err := keyExt.ReadUint32(); err != nil {
+		return err
+	}
+
+	// read out index of the milestone which contained this receipt
+	msIndex, err := keyExt.ReadUint32()
+	if err != nil {
+		return err
+	}
+
+	r := &iotago.Receipt{}
+	if _, err := r.Deserialize(value, iotago.DeSeriModeNoValidation); err != nil {
+		return err
+	}
+
+	rt.Receipt = r
+	rt.MilestoneIndex = milestone.Index(msIndex)
+
+	return nil
+}
+
+// StoreReceipt stores the given receipt tuple.
+func (u *Manager) StoreReceipt(rt *ReceiptTuple) error {
+	return u.utxoStorage.Set(rt.kvStorableKey(), rt.kvStorableValue())
+}
+
+// DeleteReceipt deletes the given receipt tuple.
+func (u *Manager) DeleteReceipt(rt *ReceiptTuple) error {
+	return u.utxoStorage.Delete(rt.kvStorableKey())
+}
+
+// adds a receipt store instruction to the given mutations.
+func storeReceipt(rt *ReceiptTuple, mutations kvstore.BatchedMutations) error {
+	return mutations.Set(rt.kvStorableKey(), rt.kvStorableValue())
+}
+
+// adds a receipt delete instruction to the given mutations.
+func deleteReceipt(rt *ReceiptTuple, mutations kvstore.BatchedMutations) error {
+	return mutations.Delete(rt.kvStorableKey())
+}
+
+// ReceiptTupleConsumer is a function that consumes a receipt tuple.
+type ReceiptTupleConsumer func(rt *ReceiptTuple) bool
+
+// ForEachReceiptTuple iterates over all stored receipt tuples.
+func (u *Manager) ForEachReceiptTuple(consumer ReceiptTupleConsumer, options ...UTXOIterateOption) error {
+	opt := iterateOptions(options)
+
```

Comment:
do we want to add this here to be consistent?
```
	if opt.readLockLedger {
		u.ReadLockLedger()
		defer u.ReadUnlockLedger()
	}
```

Same with `ForEachMigratedAtReceiptTuple`